libloragw/src/loragw_gps.c

```diff
@@ -415,88 +658,105 @@ enum gps_msg lgw_parse_ubx(const char *serial_buff, size_t buff_size, size_t *ms
     }
     DEBUG_MSG("\n");
 
+    struct cursor csr = cursor_new((char *)serial_buff, buff_size);
+
+    uint16_t ubx_sync = 0;
+    CSR_RES(cursor_unpack_be_u16(&csr, &ubx_sync));
+
     /* Check for UBX sync chars 0xB5 0x62 */
-    if ((serial_buff[0] == (char)0xB5) && (serial_buff[1] == (char)0x62)) {
-
-        /* Get payload length to compute message size */
-        payload_length  = (uint8_t)serial_buff[4];
-        payload_length |= (uint8_t)serial_buff[5] << 8;
-        *msg_size = 6 + payload_length + 2; /* header + payload + checksum */
-
-        /* check for complete message in buffer */
-        if(*msg_size <= buff_size) {
-            /* Validate checksum of message */
-            ck_a_rcv = serial_buff[*msg_size-2]; /* received checksum */
-            ck_b_rcv = serial_buff[*msg_size-1]; /* received checksum */
-            /* Use 8-bit Fletcher Algorithm to compute checksum of actual payload */
-            ck_a = 0; ck_b = 0;
-            for (i=0; i<(4 + payload_length); i++) {
-                ck_a = ck_a + serial_buff[i+2];
-                ck_b = ck_b + ck_a;
-            }
+    if (ubx_sync != 0xB562) {
+        return IGNORED;
```

Comment:
What happens if the serial output is misaligned, will it eventually resynchronize with the start of a UBX frame?

Reply:
I do occasionally see it get out of sync, and it does resynchronize fairly quickly. It's complicated by the fact that this code is processing a mixed stream of binary UBX messages and ASCII NMEA sentences. I currently haven't improved or hurt the situation, and will probably replace the whole thing with a state machine eventually.